kotlin/display-device-location/src/main/java/com/esri/arcgisruntime/sample/displaydevicelocation/MainActivity.kt

```diff
@@ -0,0 +1,159 @@
+  private val reqPermissions = arrayOf(
+    Manifest.permission.ACCESS_FINE_LOCATION,
+    Manifest.permission.ACCESS_COARSE_LOCATION
+  )
+
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    setContentView(R.layout.activity_main)
+    mapView.map = ArcGISMap(Basemap.createImagery())
+    // Listen to changes in the status of the location data source.
+    locationDisplay.addDataSourceStatusChangedListener { checkDataSource(it) }
+    // Populate the list for the Location display options for the spinner's Adapter
+    val list = arrayListOf(
+      ItemData("Stop", R.drawable.locationdisplaydisabled),
+      ItemData("On", R.drawable.locationdisplayon),
+      ItemData("Re-Center", R.drawable.locationdisplayrecenter),
+      ItemData("Navigation", R.drawable.locationdisplaynavigation),
+      ItemData("Compass", R.drawable.locationdisplayheading)
+    )
+
+    spinner.adapter = SpinnerAdapter(this, R.layout.spinner_layout, R.id.txt, list)
+    spinner.onItemSelectedListener = object : OnItemSelectedListener {
+      override fun onItemSelected(
+        parent: AdapterView<*>?,
+        view: View,
+        position: Int,
+        id: Long
+      ) {
+        when (position) {
+          0 ->  // Stop Location Display
+            if (locationDisplay.isStarted) locationDisplay.stop()
+          1 ->  // Start Location Display
+            if (!locationDisplay.isStarted) locationDisplay.startAsync()
+          2 -> {
+            // Re-Center MapView on Location
+            // AutoPanMode - Recenter: In this mode, the MapView attempts to keep the location symbol on-screen by
+            // re-centering the location symbol when the symbol moves outside a "wander extent". The location symbol
+            // may move freely within the wander extent, but as soon as the symbol exits the wander extent, the MapView
+            // re-centers the map on the symbol.
+            locationDisplay.autoPanMode = LocationDisplay.AutoPanMode.RECENTER
+            if (!locationDisplay.isStarted) locationDisplay.startAsync()
+          }
+          3 -> {
+            // Start Navigation Mode
+            // This mode is best suited for in-vehicle navigation.
+            locationDisplay.autoPanMode = LocationDisplay.AutoPanMode.NAVIGATION
+            if (!locationDisplay.isStarted) locationDisplay.startAsync()
+          }
+          4 -> {
+            // Start Compass Mode
+            // This mode is better suited for waypoint navigation when the user is walking.
+            locationDisplay.autoPanMode = LocationDisplay.AutoPanMode.COMPASS_NAVIGATION
+            if (!locationDisplay.isStarted) locationDisplay.startAsync()
+          }
+        }
+      }
+
+      override fun onNothingSelected(parent: AdapterView<*>?) {}
+    }
+  }
+
+  private fun checkDataSource(dataSourceStatusChangedEvent: LocationDisplay.DataSourceStatusChangedEvent){
```

Comment:
There's alot of code here to handle permission checks. Could it be pared down a bit towards something like this pattern: https://devtopia.esri.com/runtime/common-samples/wiki/Android-Samples-Repo%3A-A-Contribution-Guide#permissions-at-runtime---this-is-being-phased-out-with-android-10 ?

Reply:
I've tried to implement that now!